recipes/mocha-to-node-test-runner/src/workflow.ts

```diff
@@ -0,0 +1,246 @@
+import type { Edit, Range, SgRoot } from '@codemod.com/jssg-types/main';
+import isESM from '@nodejs/codemod-utils/is-esm';
+import { getNodeImportStatements } from '@nodejs/codemod-utils/ast-grep/import-statement';
+import { getNodeRequireCalls } from '@nodejs/codemod-utils/ast-grep/require-call';
+import type JS from '@codemod.com/jssg-types/langs/javascript';
+
+export default function transform(root: SgRoot<JS>): string | null {
+	const rootNode = root.root();
+
+	const globalIdentifiers = ['describe'];
+
+	const usedGlobalIdentifiers = globalIdentifiers.filter((globalIdentifier) =>
+		['', '.skip', '.only']
+			.map((suffix) => `${globalIdentifier}${suffix}($$$)`)
+			.some((pattern) => rootNode.findAll({ rule: { pattern } }).length > 0),
+	);
+
+	if (usedGlobalIdentifiers.length === 0) {
+		return null;
+	}
+
+	const edits = [
+		transformImport,
+		transformDoneCallbacks,
+		transformThisSkip,
+		transformThisTimeout,
+	].flatMap((transform) => transform(root));
+	if (!edits.length) {
+		return null;
+	}
+
+	return rootNode
+		.commitEdits(edits)
+		.split('\n')
+		.map((line) => (line.trim() === '' ? line.trim() : line))
+		.join('\n');
```

Comment:
what the goal of that ?

Reply:
When removing `this.timeout(x)` instructions, the indent remains as it's not in the AST.
Those three lines remove whitespaces in lines that only contains whitespaces.

Not sure if there is already an util to achieve it. It seems like fully removing line is still a bit painful using codemod.

Reply:
We have an utility for that on utility package called `removeLines` which allow you to pass array of `Range` .

So you can use it at:

1. intanciate an array
2. when editing pushing on this array range of thing you want to remove
2. after pushing "commit" call the function and return his result

Reply:
Doing so still leads to troubles.
- transform A edit lines (eg. 2, 6, 10)
- transform B add one more line between 3 and 4
- apply edits
- remove lines 2, 6, 10 -> oopsie, 6 and 10 are deleted instead of 7 and 11 instead.
For `mocha-to-node-test-runner` : it's gonna happen, no doubt.

I'm gonna try to implement an offset tracking for removed lines, but this is gonna bloat the code and imho we should let the formatter format things.